File: artemis-entity-tracker/src/net/namekdev/entity_tracker/network/Communicator.java
package net.namekdev.entity_tracker.network;

import java.net.SocketAddress;

import net.namekdev.entity_tracker.network.base.RawConnectionCommunicator;
import net.namekdev.entity_tracker.network.base.RawConnectionOutputListener;
import net.namekdev.entity_tracker.utils.serialization.NetworkDeserializer;
import net.namekdev.entity_tracker.utils.serialization.NetworkSerialization;
import net.namekdev.entity_tracker.utils.serialization.NetworkSerializer;

/**
 * Defines basics of network protocol for communication between EntityTracker Manager and external UI.
 *
 * @author Namek
 */
public abstract class Communicator implements RawConnectionCommunicator {
	protected RawConnectionOutputListener _output;
	protected final NetworkSerializer _serializer = NetworkSerialization.createSerializer();
	protected final NetworkDeserializer _deserializer = NetworkSerialization.createDeserializer();

	protected static final byte TYPE_ADDED_ENTITY_SYSTEM = 60;
	protected static final byte TYPE_ADDED_MANAGER = 61;
	protected static final byte TYPE_ADDED_COMPONENT_TYPE = 63;
	protected static final byte TYPE_ADDED_ENTITY = 65;
	protected static final byte TYPE_DELETED_ENTITY = 68;


	@Override
	public void connected(SocketAddress remoteAddress, RawConnectionOutputListener output) {
		_output = output;
	}

	protected void send(NetworkSerializer serializer) {
		NetworkSerializer.SerializeResult data = serializer.getResult();
		_output.send(data.buffer, 0, data.size);
	}

	protected NetworkSerializer beginPacket(byte packetType) {
		return _serializer.reset().addRawByte(packetType);
	}
}

File: src/commonMain/kotlin/net/namekdev/entity_tracker/network/Communicator.kt
package net.namekdev.entity_tracker.network

/**
 * Defines basics of network protocol for communication between EntityTracker Manager and UI.
 *
 * @author Namek
 */
abstract class Communicator : RawConnectionCommunicator {
    protected var output: RawConnectionOutputListener? = null

    override fun connected(identifier: String, output: RawConnectionOutputListener) {
        this.output = output
    }

    override fun disconnected() {}


    companion object {

        // tracker events
        const val TYPE_ADDED_SYSTEM: Byte = 60
        const val TYPE_ADDED_COMPONENT_TYPE: Byte = 63
        const val TYPE_UPDATED_SYSTEM: Byte = 64
        const val TYPE_ADDED_ENTITY: Byte = 68
        const val TYPE_DELETED_ENTITY: Byte = 73
        const val TYPE_ADDED_COMPONENT_TYPE_TO_ENTITIES: Byte = 80
        const val TYPE_REMOVED_COMPONENT_TYPE_FROM_ENTITIES: Byte = 81
        const val TYPE_UPDATED_COMPONENT_STATE: Byte = 104

        // UI requests
        const val TYPE_SET_SYSTEM_STATE: Byte = 90
        const val TYPE_DELETE_ENTITY: Byte = 100
        const val TYPE_REQUEST_COMPONENT_STATE: Byte = 103
        const val TYPE_SET_COMPONENT_FIELD_VALUE: Byte = 113
        const val TYPE_SET_COMPONENT_STATE_WATCHER: Byte = 117
    }
}
